src/analysis/retail/monk/mistweaver/modules/dragonflight/tier/T30MWTier.tsx

```diff
@@ -0,0 +1,125 @@
+import { formatNumber, formatPercentage } from 'common/format';
+import SPELLS from 'common/SPELLS';
+import RESOURCE_TYPES from 'game/RESOURCE_TYPES';
+import { TIERS } from 'game/TIERS';
+import { SpellLink } from 'interface';
+import Analyzer, { Options, SELECTED_PLAYER } from 'parser/core/Analyzer';
+import { calculateEffectiveHealing } from 'parser/core/EventCalculateLib';
+import Events, { HealEvent, ResourceChangeEvent } from 'parser/core/Events';
+import BoringValueText from 'parser/ui/BoringValueText';
+import ItemHealingDone from 'parser/ui/ItemHealingDone';
+import ItemManaGained from 'parser/ui/ItemManaGained';
+import Statistic from 'parser/ui/Statistic';
+import STATISTIC_CATEGORY from 'parser/ui/STATISTIC_CATEGORY';
+import STATISTIC_ORDER from 'parser/ui/STATISTIC_ORDER';
+
+const FOUR_PIECE_BONUS = 0.4;
+const FOUR_PIECE_SPELLS = [SPELLS.RENEWING_MIST_HEAL, SPELLS.VIVIFY];
+
+class T30TierSet extends Analyzer {
+  has2Piece: boolean = true;
+  has4Piece: boolean = true;
+  vivHealing: number = 0;
+  renewingMistHealing: number = 0;
+  manaGain: number = 0;
+  wastedManaGain: number = 0;
+
+  constructor(options: Options) {
+    super(options);
+    this.has2Piece = this.selectedCombatant.has2PieceByTier(TIERS.T29);
+    this.has4Piece = this.selectedCombatant.has4PieceByTier(TIERS.T29) && this.has2Piece;
+    this.active = this.has2Piece;
+    if (!this.active) {
+      return;
+    }
+    // 2pc do energized for mana gain value and show uptime
+    this.addEventListener(
+      Events.resourcechange.by(SELECTED_PLAYER).spell(SPELLS.SOULFANG_INFUSION),
+      this.handle2pcMana,
+    );
+    //4pc do uptime and effective healing increase
+    if (this.has4Piece) {
+      this.addEventListener(
+        Events.heal.by(SELECTED_PLAYER).spell(FOUR_PIECE_SPELLS),
+        this.handle4PcHeal,
+      );
+    }
+  }
+
+  get total4PieceHealing() {
+    return this.renewingMistHealing + this.vivHealing;
+  }
+  get twoSetUptime() {
+    return (
+      this.selectedCombatant.getBuffUptime(SPELLS.SOULFANG_INFUSION.id) / this.owner.fightDuration
+    );
+  }
+
+  get fourSetUptime() {
+    return (
+      this.selectedCombatant.getBuffUptime(SPELLS.SOULFANG_VITALITY.id) / this.owner.fightDuration
+    );
+  }
+
+  handle2pcMana(event: ResourceChangeEvent) {
+    if (event.resourceChangeType !== RESOURCE_TYPES.MANA.id) {
+      return;
+    }
+    this.manaGain += event.resourceChange - event.waste;
+    this.wastedManaGain += event.waste;
+  }
+
+  handle4PcHeal(event: HealEvent) {
+    const spellId = event.ability.guid;
+    if (this.selectedCombatant.hasBuff(SPELLS.SOULFANG_VITALITY.id)) {
+      if (spellId === SPELLS.RENEWING_MIST_HEAL.id) {
+        this.renewingMistHealing += calculateEffectiveHealing(event, FOUR_PIECE_BONUS);
+      } else if (spellId === SPELLS.VIVIFY.id) {
+        this.vivHealing += calculateEffectiveHealing(event, FOUR_PIECE_BONUS);
+      }
+    }
+  }
```

Comment:
```suggestion
  handle4PcHeal(event: HealEvent) {
    const spellId = event.ability.guid;
    if (this.selectedCombatant.hasBuff(SPELLS.SOULFANG_VITALITY.id)) {
     const increase = calculateEffectiveHealing(event, FOUR_PIECE_BONUS);
     spellId === SPELLS.RENEWING_MIST_HEAL.id ? this.renewingMistHealing += increase : this.vivifyHealing += increase;
    }
  }
```

Reply:
Not strictly necessary but I think this makes the code more concise.

Reply:
Agreed